## Data loading and merging

In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
df1 = pd.read_csv("./AA Data Challenge Datasets/prod_A.csv")

In [4]:
df1.head()

ds   userid
0  2020-01-01  3384044
1  2020-01-01  5543378
2  2020-01-01  5773857
3  2020-01-01  5526092
4  2020-01-01  6972245

In [5]:
df1.shape

(678802, 2)

In [6]:
df1.columns = ["ds1", "user_id"]

In [7]:
df2 = pd.read_csv("./AA Data Challenge Datasets/prod_B.csv")

In [8]:
df2.head()

ds   userid
0  2020-01-01  3384044
1  2020-01-01  5543378
2  2020-01-01  5773857
3  2020-01-01  5526092
4  2020-01-01  6972245

In [9]:
df2.shape

(233853, 2)

In [10]:
df2.columns = ["ds2", "user_id"]

In [11]:
df3 = pd.read_csv("./AA Data Challenge Datasets/user_data.csv")

In [12]:
df3.head()

user_id country  age  gender  start_date
0  9541124      DK   22    male  2018-05-05
1  5946065      US   15    male  2018-01-24
2  4180986      US   32  female  2018-12-17
3  5533917      US   70    male  2018-03-07
4  9531648      US   39    male  2018-09-29

In [13]:
df3.shape

(600000, 5)

In [14]:
temp_merged = pd.merge(df3, df1, left_on='user_id', right_on='user_id', how='left')

In [15]:
temp_merged.head()

user_id country  age  gender  start_date         ds1
0  9541124      DK   22    male  2018-05-05         NaN
1  5946065      US   15    male  2018-01-24  2020-03-08
2  4180986      US   32  female  2018-12-17         NaN
3  5533917      US   70    male  2018-03-07  2020-01-11
4  5533917      US   70    male  2018-03-07  2020-02-13

In [16]:
df_merged = pd.merge(temp_merged, df2, left_on='user_id', right_on='user_id', how='left')

In [17]:
df_merged.head()

user_id country  age  gender  start_date         ds1         ds2
0  9541124      DK   22    male  2018-05-05         NaN  2020-10-25
1  5946065      US   15    male  2018-01-24  2020-03-08         NaN
2  4180986      US   32  female  2018-12-17         NaN         NaN
3  5533917      US   70    male  2018-03-07  2020-01-11  2020-08-20
4  5533917      US   70    male  2018-03-07  2020-02-13  2020-08-20

In [18]:
df_merged.isna().sum() / len(df_merged) * 100

user_id        0.000000
country        0.000000
age            0.000000
gender         0.000000
start_date     0.000000
ds1           25.863805
ds2           57.617636
dtype: float64

In [20]:
df_merged.to_csv("./merged_dataset.csv", index = False)

## Average user question

In [23]:
df_merged.head()

user_id country  age  gender  start_date         ds1         ds2
0  9541124      DK   22    male  2018-05-05         NaN  2020-10-25
1  5946065      US   15    male  2018-01-24  2020-03-08         NaN
2  4180986      US   32  female  2018-12-17         NaN         NaN
3  5533917      US   70    male  2018-03-07  2020-01-11  2020-08-20
4  5533917      US   70    male  2018-03-07  2020-02-13  2020-08-20

In [24]:
df_merged["country"].value_counts()

US    577205
PL     91992
NL     84406
DE     62791
UK     49375
FR     43605
DK     42494
CZ     38173
IT     23883
FI     22064
ES      3685
Name: country, dtype: int64

In [26]:
np.average(df_merged["age"])

38.461664388706836

In [27]:
df_merged["gender"].value_counts()

male       699466
female     272324
unknown     67883
Name: gender, dtype: int64

In [30]:
df_merged["start_date"].value_counts()

2018-10-26    3154
2018-05-05    3154
2018-12-09    3153
2018-03-30    3141
2018-10-27    3139
              ... 
2018-11-09    2715
2018-03-21    2700
2018-11-03    2693
2018-08-29    2681
2018-11-10    2678
Name: start_date, Length: 356, dtype: int64

In [33]:
pd.DatetimeIndex(df_merged['start_date']).year.value_counts()

2018    1039673
Name: start_date, dtype: int64

In [34]:
pd.DatetimeIndex(df_merged['start_date']).month.value_counts()

5     91297
1     91076
3     90800
10    90627
7     90196
8     90124
6     87697
4     87377
9     87236
11    86030
2     82482
12    64731
Name: start_date, dtype: int64

In [35]:
pd.DatetimeIndex(df_merged['start_date']).day.value_counts()

9     35548
18    35394
16    35385
21    35276
13    35183
7     35064
22    35025
5     35024
12    35020
19    35007
14    35007
17    34981
10    34974
11    34960
15    34924
20    34870
1     34855
8     34779
3     34776
4     34695
6     34693
2     34579
23    32560
27    32528
26    32447
28    32325
24    32210
25    31868
30    29312
29    28840
31    17564
Name: start_date, dtype: int64

### **Average user:**
- **Country** : US
- **Age** : ~38.5
- **Gender** : Male
- **Start date** : 09 May 2018

## Question 2

In [36]:
df_merged.head()

user_id country  age  gender  start_date         ds1         ds2
0  9541124      DK   22    male  2018-05-05         NaN  2020-10-25
1  5946065      US   15    male  2018-01-24  2020-03-08         NaN
2  4180986      US   32  female  2018-12-17         NaN         NaN
3  5533917      US   70    male  2018-03-07  2020-01-11  2020-08-20
4  5533917      US   70    male  2018-03-07  2020-02-13  2020-08-20

In [45]:
product1 = df_merged.dropna(subset=["ds1"]).reset_index().drop(["ds2"], axis = 1)

In [46]:
product2 = df_merged.dropna(subset=["ds2"]).reset_index().drop(["ds1"], axis = 1)

In [47]:
product1.head()

index  user_id country  age gender  start_date         ds1
0      1  5946065      US   15   male  2018-01-24  2020-03-08
1      3  5533917      US   70   male  2018-03-07  2020-01-11
2      4  5533917      US   70   male  2018-03-07  2020-02-13
3      5  5533917      US   70   male  2018-03-07  2020-06-25
4      6  5533917      US   70   male  2018-03-07  2020-06-28

In [48]:
product2.head()

index  user_id country  age gender  start_date         ds2
0      0  9541124      DK   22   male  2018-05-05  2020-10-25
1      3  5533917      US   70   male  2018-03-07  2020-08-20
2      4  5533917      US   70   male  2018-03-07  2020-08-20
3      5  5533917      US   70   male  2018-03-07  2020-08-20
4      6  5533917      US   70   male  2018-03-07  2020-08-20

#### Product 1 analysis

In [49]:
product1["country"].value_counts()

US    368624
PL     79935
NL     73357
DE     54553
UK     42927
FR     37899
DK     37074
CZ     33309
IT     20720
FI     19141
ES      3235
Name: country, dtype: int64

In [50]:
np.average(product1["age"])

42.61253493241858

In [58]:
product1["gender"].value_counts()

male       518454
female     201924
unknown     50396
Name: gender, dtype: int64

In [62]:
pd.DatetimeIndex(product1['start_date']).year.value_counts()

2018    770774
Name: start_date, dtype: int64

In [63]:
pd.DatetimeIndex(product1['start_date']).month.value_counts()

5     67641
1     67488
10    67346
3     67305
7     67260
8     66620
6     65137
4     64834
9     64250
11    63600
2     61130
12    48163
Name: start_date, dtype: int64

In [64]:
pd.DatetimeIndex(product1['start_date']).day.value_counts()

9     26533
16    26392
18    26392
21    26314
7     26067
13    26063
14    26036
5     25986
12    25936
19    25864
15    25832
3     25827
22    25820
10    25817
11    25814
8     25804
1     25779
17    25766
20    25714
4     25706
2     25676
6     25605
27    24259
23    24242
26    24186
28    23989
24    23904
25    23561
30    21776
29    21064
31    13050
Name: start_date, dtype: int64

#### Product 2 analysis

In [55]:
product2["country"].value_counts()

US    171241
PL     53513
NL     49113
DE     36760
UK     28849
FR     25330
DK     24817
CZ     22051
IT     13831
FI     12929
ES      2204
Name: country, dtype: int64

In [56]:
np.average(product2["age"])

41.28755123253101

In [57]:
product2["gender"].value_counts()

male       296301
female     115596
unknown     28741
Name: gender, dtype: int64

In [59]:
pd.DatetimeIndex(product2['start_date']).year.value_counts()

2018    440638
Name: start_date, dtype: int64

In [60]:
pd.DatetimeIndex(product2['start_date']).month.value_counts()

5     38951
10    38760
1     38539
3     38449
7     38298
8     38215
6     37101
4     36873
9     36823
11    36305
2     35299
12    27025
Name: start_date, dtype: int64

In [61]:
pd.DatetimeIndex(product2['start_date']).day.value_counts()

9     15324
21    15093
10    15063
16    15005
5     14950
14    14933
12    14918
18    14918
4     14896
22    14890
3     14838
19    14830
20    14798
15    14756
2     14747
17    14720
13    14708
7     14650
8     14636
1     14555
6     14508
11    14436
23    14110
28    14011
27    13750
24    13745
26    13717
25    13402
30    12667
29    11855
31     7209
Name: start_date, dtype: int64